In [1]:
import pandas as pd

In [20]:
input_path = r"data/staySpain_cleaned.pkl"
df = pd.read_pickle(input_path)

## PREGUNTA NEGOCI

### preu mitjà per tipus d'allotjament a cada ciutat

In [24]:
# Create crosstab with mean price
mean_price_per_room_type_and_city = pd.crosstab(index=df['city'], columns=df['room_type'], values=df['price'], aggfunc='mean')

# Display the crosstab
display(round(mean_price_per_room_type_and_city,2))

room_type,Entire home/apt,Hotel room,Private room,Shared room
city,,,,
barcelona,1300.94,1444.29,455.05,371.88
girona,1230.68,625.00,602.12,NaN
madrid,942.63,902.50,372.81,313.33
malaga,815.78,703.33,356.73,143.33
mallorca,1709.57,1401.67,700.25,493.33
menorca,1656.53,NaN,610.00,NaN
sevilla,1033.91,918.57,606.00,120.00
valencia,796.62,955.00,340.15,203.33


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6733 entries, 0 to 6999
Data columns (total 35 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   apartment_id                 6733 non-null   int64         
 1   name                         6730 non-null   object        
 2   description                  6706 non-null   object        
 3   host_id                      6733 non-null   int64         
 4   neighbourhood_name           6733 non-null   object        
 5   neighbourhood_district       4075 non-null   object        
 6   room_type                    6733 non-null   object        
 7   accommodates                 6733 non-null   int64         
 8   bathrooms                    6701 non-null   Int64         
 9   bedrooms                     6705 non-null   Int64         
 10  beds                         6730 non-null   float64       
 11  amenities_list               6717 non-null   obj

## NULOS

In [14]:
# Create crosstab for null counts in 'price'
crosstab_nulls = pd.crosstab(df['city'], df['room_type'], values=df['price'].isnull(), aggfunc='sum')

display(crosstab_nulls)

room_type,Entire home/apt,Hotel room,Private room,Shared room
city,,,,
barcelona,21.0,0.0,3.0,2.0
girona,15.0,0.0,0.0,NaN
madrid,8.0,1.0,1.0,0.0
malaga,1.0,0.0,0.0,0.0
mallorca,58.0,1.0,1.0,0.0
menorca,4.0,NaN,0.0,NaN
sevilla,1.0,0.0,1.0,0.0
valencia,2.0,0.0,1.0,0.0


In [ ]:
En valores absolutos,
Los nulos están concentrados en
MALLORCA - Entire home/apt : 58   (5.76%)
BARCELONA - Entire home/apt : 21  (2.21%)
GIRONA - Entire home/apt: 15   (1.45%)


In [12]:
# Create crosstab for total counts
crosstab_total = pd.crosstab(df['city'], df['room_type'])

display(crosstab_total)

room_type,Entire home/apt,Hotel room,Private room,Shared room
city,,,,
barcelona,950,7,1066,18
girona,1029,4,52,0
madrid,901,5,475,15
malaga,278,3,55,3
mallorca,1006,7,80,3
menorca,128,0,10,0
sevilla,262,7,71,1
valencia,226,2,66,3


In [13]:
# Calculate the proportion of nulls
proportion_nulls = crosstab_nulls.div(crosstab_total).fillna(0)

display(proportion_nulls*100) # shows percentage

room_type,Entire home/apt,Hotel room,Private room,Shared room
city,,,,
barcelona,2.210526,0.000000,0.281426,11.111111
girona,1.457726,0.000000,0.000000,0.000000
madrid,0.887902,20.000000,0.210526,0.000000
malaga,0.359712,0.000000,0.000000,0.000000
mallorca,5.765408,14.285714,1.250000,0.000000
menorca,3.125000,0.000000,0.000000,0.000000
sevilla,0.381679,0.000000,1.408451,0.000000
valencia,0.884956,0.000000,1.515152,0.000000


Hotel Room - Madrid (20%)
Hotel Room - Mallorca (14.28%)
Sharted Room - Barcelona (11.11% PERO hay muy poca muestra)

In [25]:
nulos_df = df[df.price.isna()]

In [19]:
nulos_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 121 entries, 16 to 6993
Data columns (total 35 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   apartment_id                 121 non-null    int64         
 1   name                         121 non-null    object        
 2   description                  120 non-null    object        
 3   host_id                      121 non-null    int64         
 4   neighbourhood_name           121 non-null    object        
 5   neighbourhood_district       41 non-null     object        
 6   room_type                    121 non-null    object        
 7   accommodates                 121 non-null    int64         
 8   bathrooms                    121 non-null    Int64         
 9   bedrooms                     121 non-null    Int64         
 10  beds                         120 non-null    float64       
 11  amenities_list               119 non-null    obj

In [26]:
pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns

In [27]:
nulos_df

,apartment_id,name,description,host_id,neighbourhood_name,neighbourhood_district,room_type,accommodates,bathrooms,bedrooms,beds,amenities_list,price,minimum_nights,maximum_nights,has_availability,availability_30,availability_60,availability_90,availability_365,number_of_reviews,first_review_date,last_review_date,review_scores_rating,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,is_instant_bookable,reviews_per_month,country,city,insert_date
16,101580,Big house in PARADISE,Luxury and quiet villa in Cala Canyelles (2km ...,532063,Lloret de Mar,NaN,Entire home/apt,16,4,4,12.0,"TV,Cable TV,Internet,Wifi,Air conditioning,Poo...",NaN,2,30,VERDADERO,22,52,82,237,14,2017-05-28,2018-10-14,99.0,10.0,10.0,10.0,10.0,9.0,10.0,VERDADERO,73.0,spain,girona,2018-12-22
55,361488,RETIRO PARK2 CHALET INDUSTRIAL-CHIC,CHALET INDUSTRIAL-CHIC RETIRO PARK2 This vill...,1732442,Ni�o Jes�s,Retiro,Entire home/apt,6,3,3,3.0,"TV,Internet,Wifi,Kitchen,Free parking on premi...",NaN,5,365,VERDADERO,0,0,0,0,0,NaT,NaT,NaN,NaN,NaN,NaN,NaN,NaN,NaN,FALSO,NaN,spain,madrid,2019-04-11
85,474387,"Des Artistes 716, wifi, central P",THE PRICE IS FOR EXCLUSIVE USE FOR A MAX. OF 1...,1614471,el Fort Pienc,Eixample,Entire home/apt,11,2,5,11.0,"Wireless Internet,Wheelchair accessible,Smokin...",NaN,3,30,NaN,8,38,68,340,16,2013-04-22,2016-02-25,81.0,9.0,9.0,10.0,9.0,9.0,8.0,FALSO,35.0,spain,barcelona,2017-02-09
160,721378,Big and nice apartment near to Sagrada Familia,Beautiful 4 bedroom apartment in a superb posi...,3724467,el Fort Pienc,Eixample,Entire home/apt,6,2,4,5.0,"TV,Internet,Wifi,Air conditioning,Kitchen,Paid...",NaN,3,365,VERDADERO,15,35,65,340,3,2018-03-03,2019-02-27,93.0,10.0,9.0,10.0,8.0,9.0,9.0,VERDADERO,16.0,spain,barcelona,2019-09-17
161,721509,P3 Valencia Apartment 1 bedroom,"Lovely one-bedroom apartment, located on a fir...",3724467,la Dreta de l'Eixample,Eixample,Entire home/apt,4,1,1,1.0,"TV,Cable TV,Internet,Wifi,Air conditioning,Kit...",NaN,3,365,VERDADERO,23,48,78,158,1,2018-04-18,2018-04-18,80.0,6.0,10.0,10.0,10.0,8.0,6.0,VERDADERO,10.0,spain,barcelona,2018-05-14
308,1071662,LUXURY SEAVIEW S'AGARO COSTA BRAVA,BEST HOLIDAY ON THE COSTA BRAVA THAT YOU CAN T...,1830359,Castell-Platja d'Aro,NaN,Entire home/apt,7,2,3,6.0,"TV,Wifi,Air conditioning,Pool,Kitchen,Free par...",NaN,14,60,VERDADERO,23,53,83,358,0,NaT,NaT,NaN,NaN,NaN,NaN,NaN,NaN,NaN,FALSO,NaN,spain,girona,2018-12-22
476,1648617,Holiday House in CALA TUENT,Historic finca with private access to the sea....,8746555,Escorca,NaN,Entire home/apt,12,5,6,12.0,"TV,Cable TV,Wifi,Pool,Kitchen,Free parking on ...",NaN,3,1125,VERDADERO,6,15,24,291,38,2014-07-14,2019-06-10,99.0,10.0,10.0,10.0,10.0,10.0,10.0,FALSO,63.0,spain,mallorca,2019-06-27
633,2531084,In front of the Giralda,"This apartment, situated in �Calle Alemanes�, ...",4653961,Santa Cruz,Casco Antiguo,Entire home/apt,4,1,1,2.0,"TV,Internet,Wifi,Air conditioning,Kitchen,Paid...",NaN,3,1124,VERDADERO,20,48,66,307,57,2014-05-30,2018-07-27,97.0,10.0,10.0,10.0,10.0,10.0,10.0,VERDADERO,112.0,spain,sevilla,2018-07-31
707,2963691,Modern villa in Portals with pool and sea view,"Only 5 min to the nearest beach, to the tennis...",15124488,Calvi�,NaN,Entire home/apt,8,4,4,8.0,"TV,Internet,Wifi,Air conditioning,Pool,Kitchen...",NaN,4,1125,VERDADERO,30,58,61,332,27,2015-08-19,2019-08-31,99.0,10.0,10.0,10.0,10.0,10.0,9.0,FALSO,46.0,spain,mallorca,2020-05-28
743,3107710,1563 - AB Pla�a Espanya 2-1 - Classy 3-Bedroom...,This modish three bedroom holiday apartment in...,4459553,la Nova Esquerra de l'Eixample,Eixample,Entire home/apt,6,2,3,6.0,"TV,Internet,Wifi,Air conditioning,Kitchen,Elev...",NaN,2,330,VERDADERO,9,34,64,339,11,2016-06-20,2018-08-20,80.0,8.0,9.0,8.0,8.0,9.0,8.0,VERDADERO,38.0,spain,barcelona,2018-11-07
